File: web/src/repositories/eventRepository.ts
import type { HealthEvent, NewHealthEvent } from "@/domain/healthEvent";
import type {
  EventRepository,
  InsertRawRecordResult,
} from "@/domain/repositories";
import type { NewRawRecord, RawRecord } from "@/domain/rawRecord";
import type { Json } from "./supabase/databaseTypes";
import { createSupabaseServerClient } from "./supabase/serverClient";

const UNIQUE_VIOLATION = "23505";

function toRawRecord(row: {
  id: number;
  source: string;
  record_type: string;
  external_id: string | null;
  payload: unknown;
  payload_hash: string;
  device_id: string | null;
  received_at: string;
  norm_status: string;
  norm_error: string | null;
}): RawRecord {
  return {
    id: row.id,
    source: row.source as RawRecord["source"],
    recordType: row.record_type,
    externalId: row.external_id,
    payload: row.payload as Record<string, unknown>,
    payloadHash: row.payload_hash,
    deviceId: row.device_id,
    receivedAt: row.received_at,
    normStatus: row.norm_status as RawRecord["normStatus"],
    normError: row.norm_error,
  };
}

function toHealthEvent(row: {
  id: number;
  event_type: string;
  start_time: string;
  end_time: string | null;
  value: number | null;
  unit: string | null;
  detail: unknown;
  source: string;
  raw_record_id: number | null;
  superseded_by: number | null;
  created_at: string;
}): HealthEvent {
  return {
    id: row.id,
    eventType: row.event_type as HealthEvent["eventType"],
    startTime: row.start_time,
    endTime: row.end_time,
    value: row.value,
    unit: row.unit,
    detail: row.detail as Record<string, unknown> | null,
    source: row.source,
    rawRecordId: row.raw_record_id,
    supersededBy: row.superseded_by,
    createdAt: row.created_at,
  };
}

export async function createSupabaseEventRepository(): Promise<EventRepository> {
  const supabase = await createSupabaseServerClient();

  return {
    async insertRawRecord(
      record: NewRawRecord,
    ): Promise<InsertRawRecordResult> {
      const { data, error } = await supabase
        .from("raw_records")
        .insert({
          source: record.source,
          record_type: record.recordType,
          external_id: record.externalId,
          payload: record.payload as Json,
          payload_hash: record.payloadHash,
          device_id: record.deviceId,
        })
        .select()
        .single();

      if (error) {
        if (error.code === UNIQUE_VIOLATION) {
          return { status: "duplicate" };
        }
        throw error;
      }

      return { status: "accepted", record: toRawRecord(data) };
    },

    async listPendingRawRecords(limit = 100): Promise<RawRecord[]> {
      const { data, error } = await supabase
        .from("raw_records")
        .select()
        .eq("norm_status", "pending")
        .order("received_at", { ascending: true })
        .limit(limit);

      if (error) throw error;
      return data.map(toRawRecord);
    },

    async markRawRecordNormalized(id, result) {
      const { error } = await supabase
        .from("raw_records")
        .update(
          result.status === "done"
            ? { norm_status: "done", norm_error: null }
            : { norm_status: "error", norm_error: result.error },
        )
        .eq("id", id);

      if (error) throw error;
    },

    async insertHealthEvents(
      events: NewHealthEvent[],
    ): Promise<HealthEvent[]> {
      if (events.length === 0) return [];

      const { data, error } = await supabase
        .from("health_events")
        .insert(
          events.map((event) => ({
            event_type: event.eventType,
            start_time: event.startTime,
            end_time: event.endTime,
            value: event.value,
            unit: event.unit,
            detail: event.detail as Json | null,
            source: event.source,
            raw_record_id: event.rawRecordId,
          })),
        )
        .select();

      if (error) throw error;
      return data.map(toHealthEvent);
    },

    async listHealthEvents(params): Promise<HealthEvent[]> {
      let query = supabase
        .from("health_events")
        .select()
        .is("superseded_by", null)
        .order("start_time", { ascending: false });

      if (params.eventType) query = query.eq("event_type", params.eventType);
      if (params.from) query = query.gte("start_time", params.from);
      if (params.to) query = query.lte("start_time", params.to);

      const { data, error } = await query;
      if (error) throw error;
      return data.map(toHealthEvent);
    },
  };
}

File: server/app/repositories/sqlite/event_repository.py
import json
from datetime import UTC, datetime

from sqlalchemy import text
from sqlalchemy.engine import Engine

from app.domain.models import HealthEvent


class SqliteEventRepository:
    """Implementação SQLite de EventRepository (app.domain.repositories)."""

    def __init__(self, engine: Engine) -> None:
        self._engine = engine

    def add(self, event: HealthEvent) -> HealthEvent:
        now = datetime.now(UTC).isoformat()
        with self._engine.begin() as conn:
            result = conn.execute(
                text(
                    """
                    INSERT INTO health_events
                        (event_type, start_time, end_time, value, unit, detail,
                         source, raw_record_id, created_at)
                    VALUES
                        (:event_type, :start_time, :end_time, :value, :unit, :detail,
                         :source, :raw_record_id, :created_at)
                    """
                ),
                {
                    "event_type": event.event_type,
                    "start_time": event.start_time.isoformat(),
                    "end_time": event.end_time.isoformat() if event.end_time else None,
                    "value": event.value,
                    "unit": event.unit,
                    "detail": json.dumps(event.detail) if event.detail is not None else None,
                    "source": event.source,
                    "raw_record_id": event.raw_record_id,
                    "created_at": now,
                },
            )
            new_id = result.lastrowid
        return event.model_copy(update={"id": new_id, "created_at": datetime.fromisoformat(now)})

    def list_by_type(self, event_type: str, start: datetime, end: datetime) -> list[HealthEvent]:
        with self._engine.connect() as conn:
            rows = conn.execute(
                text(
                    """
                    SELECT id, event_type, start_time, end_time, value, unit, detail,
                           source, raw_record_id, superseded_by, created_at
                    FROM health_events
                    WHERE event_type = :event_type
                      AND start_time >= :start
                      AND start_time < :end
                      AND superseded_by IS NULL
                    ORDER BY start_time
                    """
                ),
                {
                    "event_type": event_type,
                    "start": start.isoformat(),
                    "end": end.isoformat(),
                },
            ).fetchall()
        return [self._row_to_event(row) for row in rows]

    @staticmethod
    def _row_to_event(row) -> HealthEvent:
        return HealthEvent(
            id=row.id,
            event_type=row.event_type,
            start_time=datetime.fromisoformat(row.start_time),
            end_time=datetime.fromisoformat(row.end_time) if row.end_time else None,
            value=row.value,
            unit=row.unit,
            detail=json.loads(row.detail) if row.detail else None,
            source=row.source,
            raw_record_id=row.raw_record_id,
            superseded_by=row.superseded_by,
            created_at=datetime.fromisoformat(row.created_at),
        )
